### PR TITLE
Fix expiration in RoleCache

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix expiration in `RoleCache`.

`RoleCache` is a cache containing information about roles and access rights extracted from roles. Sometimes it could work incorrectly after a role in the cache expired and then got changed by admin granting or revoking something. This PR fixes that problem.

The problem in the code which this PR is fixing was found by @ilejn in https://github.com/ClickHouse/ClickHouse/pull/66845

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 4e548a287bc26ae5814bdc17105ecf5e7ee3123c with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67748/4e548a287bc26ae5814bdc17105ecf5e7ee3123c/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/e4c165e919df7b43671fd8909176e2162c9a65fc/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: tests/integration/test_role/test.py [hunk 1]
```diff
@@ -487,3 +488,158 @@ def test_two_roles_expiration():
     instance.query("DROP ROLE rre_second")
     instance.query("DROP TABLE tre")
     instance.query("DROP TABLE tre1")
+
+
+def test_roles_cache():
```
Comment (vitlibar) [inline tests/integration/test_role/test.py:493]: This test is not simple, but it really reproduces the issue quite often.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by ilejn** [comment]
@vitlibar , thank you for this PR and for the test, it should help many CH users.

But could you please explain why the code is necessary ?

In other words, why collectEnabledRoles is not enough, although it refreshes everything and does it in 'transaction' way because of mutex in AccessChangesNotifier?

**Comment by ilejn** [comment]
> why collectEnabledRoles is not enough, although it refreshes everything and does it in 'transactional' way because of mutex in AccessChangesNotifier?

For the record - collectEnabledRoles can be called from getEnabledRoles. In this case it refreshes what is related to a user, not definitely all roles.
The cache is global, so during getting  information about a user, a role, that is not granted to the user, can expire.

File: tests/integration/test_role/test.py [hunk 2]
```diff
@@ -418,72 +419,215 @@ def test_function_current_roles():
     )
 
 
-def test_role_expiration():
-    instance.query("CREATE USER ure")
+@pytest.mark.parametrize("with_extra_role", [False, True])
+def test_role_expiration(with_extra_role):
```
Comment (vitlibar) [inline tests/integration/test_role/test.py:423]: Here I combined tests `test_role_expiration` and `test_two_roles_expiration` together.

File: tests/integration/test_role/test.py [hunk 3]
```diff
@@ -418,72 +419,215 @@ def test_function_current_roles():
...
 
-    instance.query("CREATE TABLE IF NOT EXISTS tre1 (id Int) Engine=Log")
-    instance.query("INSERT INTO tre1 VALUES (0)")
-    instance.query("GRANT SELECT ON tre1 TO rre")
+    if with_extra_role:
+        # Expiration of role "rre" from the role cache can be caused by another role being used.
+        instance.query("CREATE ROLE extra_role")
+        instance.query("CREATE USER extra_user DEFAULT ROLE extra_role")
+        instance.query("GRANT SELECT ON table1 TO extra_role")
+        assert instance.query("SELECT * FROM table1", user="extra_user") == "1\n"
```
Comment (vitlibar) [inline tests/integration/test_role/test.py:448]: Line `instance.query("SELECT * FROM table1", user="extra_user")` was missing in test `test_two_roles_expiration` and that's why test `test_two_roles_expiration` didn't fail with that bug (https://github.com/ClickHouse/ClickHouse/pull/66845, see "1.1. new_subscriptions_on_roles are intact here") in the code.

**Comment by vitlibar** [comment]
> But could you please explain why the code is necessary ?
> 
> In other words, why collectEnabledRoles is not enough, although it refreshes everything and does it in 'transactional' way because of mutex in AccessChangesNotifier?

That is explained in the comment:
https://github.com/ClickHouse/ClickHouse/blob/7ef5c366e873c4fd99f257eefbb3a350848e308c/src/Access/RoleCache.h#L43-L44

**Comment by vitlibar** [comment]
CI failures are unrelated:
- [test_ssl_cert_authentication/test.py::test_native_wrong_cert](https://s3.amazonaws.com/clickhouse-test-reports/67748/7ef5c366e873c4fd99f257eefbb3a350848e308c/integration_tests__aarch64__[6_6].html) - flaky test, see https://github.com/ClickHouse/ClickHouse/issues/68254

**Comment by tavplubix** [comment]
The new test is flaky:
https://s3.amazonaws.com/clickhouse-test-reports/0/f7ed0912db96f884a40bbcdaedfd4128cc70cb57/integration_tests__asan__old_analyzer__[5_6].html
https://s3.amazonaws.com/clickhouse-test-reports/0/34c14a6493ef4abb141b7aa4eeaa6d9b9cdc39d4/integration_tests__aarch64__[5_6].html

**Comment by ilejn** [comment]
> The new test is flaky:

@tavplubix ,  see https://github.com/ClickHouse/ClickHouse/pull/69308

**Comment by boldakovds** [comment]
Hi! I can't find version 24.3.(?) that was fixed with backport. Help me find out this version please

**Comment by ilejn** [comment]
> Hi! I can't find version 24.3.(?) that was fixed with backport. Help me find out this version please

As far as I can see it is in 24.8 branch and should be released soon, but not yet.

**Pull request merged**